File: v2_typescript/src/btc_slug.ts
const WINDOW_15M_SEC = 900;

export function getSlug15m(useNextWindow: boolean): string {
  const now = Math.floor(Date.now() / 1000);
  const windowStart = Math.floor(now / WINDOW_15M_SEC) * WINDOW_15M_SEC;
  const start = useNextWindow ? windowStart + WINDOW_15M_SEC : windowStart;
  return `btc-updown-15m-${start}`;
}

export function getWindowEndTs15m(): number {
  const now = Math.floor(Date.now() / 1000);
  const start = Math.floor(now / WINDOW_15M_SEC) * WINDOW_15M_SEC;
  return start + WINDOW_15M_SEC;
}

export function getTimeRemainingSec15m(): number {
  const now = Date.now() / 1000;
  const end = getWindowEndTs15m();
  return Math.max(0, end - now);
}

File: v2_rust/src/btc_slug.rs
//! BTC up/down market slug. 15m: btc-updown-15m-{unix_window_start}

pub const WINDOW_15M_SEC: u64 = 900;

pub fn get_slug_15m(use_next_window: bool) -> String {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    let window_start = (now / WINDOW_15M_SEC) * WINDOW_15M_SEC;
    let start = if use_next_window {
        window_start + WINDOW_15M_SEC
    } else {
        window_start
    };
    format!("btc-updown-15m-{}", start)
}

pub fn get_window_end_ts_15m() -> u64 {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    let start = (now / WINDOW_15M_SEC) * WINDOW_15M_SEC;
    start + WINDOW_15M_SEC
}

pub fn get_time_remaining_sec_15m() -> f64 {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64();
    let end = get_window_end_ts_15m() as f64;
    (end - now).max(0.0)
}
